lib/future.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+class Future {
+  constructor(executor) {
+    this.executor = executor;
+  }
+
+  static of(value) {
+    return new Future(resolve => resolve(value));
+  }
+
+  static err(error) {
+    return new Future((resolve, reject) => reject(error));
+  }
+
+  chain(fn) {
+    return new Future((resolve, reject) =>
+      this.run(value => fn(value).run(resolve, reject), error => reject(error))
+    );
+  }
+
+  map(fn) {
+    return this.chain(
+      value =>
+        new Future((resolve, reject) => {
+          try {
+            resolve(fn(value));
+          } catch (error) {
+            reject(error);
+          }
+        })
+    );
+  }
+
+  run(successed, failed) {
+    this.executor(successed, failed);
```

Comment:
I think it's important to catch errors the `executor` may throw. It will also allow omitting error handling in the `map` method